apollo-router-core/src/json_ext.rs

```diff
@@ -212,6 +132,160 @@ impl ValueExt for Value {
+    #[track_caller]
+    fn from_path(path: &Path, value: Value) -> Value {
+        let mut res_value = Value::default();
+        let mut current_node = &mut res_value;
+
+        for p in path.iter() {
+            match p {
+                PathElement::Flatten => {
+                    let a = Vec::new();
+                    *current_node = Value::Array(a);
+                }
+
+                &PathElement::Index(index) => match current_node {
+                    Value::Array(a) => {
+                        for _ in 0..index {
+                            a.push(Value::default());
+                        }
+                        a.push(Value::default());
+                        current_node = a
+                            .get_mut(index)
+                            .expect("we just created the value at that index");
+                    }
+                    Value::Null => {
+                        let mut a = Vec::new();
+                        for _ in 0..index {
+                            a.push(Value::default());
+                        }
+                        a.push(Value::default());
+
+                        *current_node = Value::Array(a);
+                        current_node = current_node
+                            .as_array_mut()
+                            .expect("current_node was just set to a Value::Array")
+                            .get_mut(index)
+                            .expect("we just created the value at that index");
+                    }
+                    other => unreachable!("unreachable node: {:?}", other),
+                },
+                PathElement::Key(k) => {
+                    let mut m = Map::new();
+                    m.insert(k.to_string(), Value::default());
+
+                    *current_node = Value::Object(m);
+                    current_node = current_node
+                        .as_object_mut()
+                        .expect("current_node was just set to a Value::Object")
+                        .get_mut(k)
+                        .expect("the value at that key was just inserted");
+                }
+            }
+        }
+
+        *current_node = value;
+        res_value
+    }
+}
+
+/// selects all values matching a Path
+///
+/// this will return the values and their specific path in the Value:
+/// a Path with a flatten node can match multiple values
+pub(crate) fn select_values_and_paths<'a>(
+    path: &'a Path,
+    data: &'a Value,
+) -> Result<Vec<(Path, &'a Value)>, FetchError> {
+    let mut res = Vec::new();
+    match iterate_path(&Path::default(), &path.0, data, &mut res) {
+        Some(err) => Err(err),
+        None => Ok(res),
+    }
+}
+
+#[cfg(test)]
+pub(crate) fn select_values<'a>(
+    path: &'a Path,
+    data: &'a Value,
+) -> Result<Vec<&'a Value>, FetchError> {
+    Ok(select_values_and_paths(path, data)?
+        .into_iter()
+        .map(|r| r.1)
+        .collect())
+}
+
+fn iterate_path<'a>(
```

Comment:
suggestion by @o0Ignition0o : using a visitor here would be great, it would help in refactoring the code in fetch_node

Reply:
I'm not sure what he meant by visitor. Did you mean iterator?

Reply:
I meant a [visitor](https://rust-unofficial.github.io/patterns/patterns/behavioural/visitor.html)

Reply:
🦎

Reply:
I added a closure based visitor in 7f7de6c